common/djangoapps/track/contexts.py

```diff
@@ -12,26 +12,27 @@
 log = logging.getLogger(__name__)
 
 
-def course_context_from_url(url):
+def course_context_from_url(url, course_id_string=None):
     """
-    Extracts the course_context from the given `url` and passes it on to
-    `course_context_from_course_id()`.
+    If course_id_string string is not present, extracts it from the given `url`. Either way, then passes
+    it on to `course_context_from_course_id()`.
     """
     url = url or ''
-
-    match = COURSE_REGEX.match(url)
     course_id = None
-    if match:
-        course_id_string = match.group('course_id')
-        try:
-            course_id = CourseKey.from_string(course_id_string)
-        except InvalidKeyError:
-            log.warning(
-                'unable to parse course_id "{course_id}"'.format(
-                    course_id=course_id_string
-                ),
-                exc_info=True
-            )
+
+    if course_id_string is None:
+        match = COURSE_REGEX.match(url)
+        if match:
+            course_id_string = match.group('course_id')
+    try:
+        course_id = CourseKey.from_string(course_id_string)
```

Comment:
We're calling the field `courserun_key`, but I think it's probably better to parse out a `LearningContextKey` here instead. `LearningContextKey` is a superset of `CourseKey`, that will also parse the keys for v2 content libraries, learning pathways, and others.

Thoughts @doctoryes?

Reply:
Or maybe we could have two fields, one with the `courserun_key` and the other explicitly a `learning_context_key`? @pwnage101: ^ thoughts on this?

Reply:
I think it makes sense to add both.

Does that mean the courserun_key would be empty if the learning context key points to something other than a courserun? FYI @brianhw @jfinkiel

Reply:
First of all, this is for writing out tracking log events, not Segment events (at least not directly).  It involves getting values into the context.course_id field of these events.   It does not propose making any change to add a courserun_key value or a learning_context_key value.  It is allowing "browser" events to explicitly pass an argument so that a known value can be put in the event instead of relying on brittle pattern-matching on the Request.path.   The purpose, as I understand it, is to reduce the risk of the pattern-matching failing when URL formats change.  The context/course_id value is used by our Event export pipeline, to route events to the corresponding Partner based on the "org" component of the context/course_id.  

So we could change from CourseKey to LearningContextKey, and I would expect little difference in the above behavior.  The context/course_id will be parsed by our pipeline with CourseKey, and will fail on those learning contexts that are not courseruns. 

For events that then get forwarded automatically to Segment, however, having the broader learning-context value in context/course_id will allow us to populate more values (in theory) when putting this value in the Segment event.  (Which currently puts it in the "label" property, but that's another story.).  I say "in theory", because I don't know if events emitted where the global $$course_id is set will ever have a non-courserun value.

Reply:
So my apologies.   I believe I was possibly mixing up two separate things.  One thing is the argument being passed into /event, which is `courserun_key`.   This is distinct from the value to which it's being written in the tracking log event, which is `context.course_id`.   It's important that the latter continue to be populated in the tracking log.  But:
* I am fine with additional values being written out to tracking logs, e.g. writing `context.courserun_key` and or `context.learning_context_key` in addition, in those cases where we know the difference.
* I defer to others on whether to have one or two parameters on the /event request, for courserun_key and learning_context_key.  I guess it's been pointed out that a learner may be operating on a problem that is pulled from a library, but doing so within the context of a real course_run, so it's possible to have both values be set?  (I confess I don't really have a good model of when learners are not within a course run.  So I would particularly look to @ormsbee to set me straight.)

Reply:
Would it be reasonable to leave this PR as-is, and consider adding new parameters (like learning_context_key) in a later issue? Specifically to avoid scope-creep...

Reply:
per followup with @ormsbee, leaving any adjustment to exactly what keys we log or what keys we get from from the frontend for later PRs.